tasks/image-region-extraction/main.py

```diff
@@ -0,0 +1,161 @@
+
+    @property
+    def witdh(self) -> int:
+        return self.tr.x - self.tl.x
+
+    @property
+    def height(self) -> int:
+        return self.bl.y - self.tl.y
+
+    def numpy(self) -> np.ndarray:
+        return np.array([
+            [self.tl.x, self.tl.y],
+            [self.tr.x, self.tr.y],
+            [self.br.x, self.br.y],
+            [self.bl.x, self.bl.y]
+        ], dtype = np.float32)
+
+    def center(self) -> tuple[int, int]:
+        temp = np.sum(self.numpy(), axis = 0) // 4
+        return temp
+
+    @classmethod
+    def sortRectPoints(cls, points: list[Point2D]) -> list[Point2D]:
+        xSorted = sorted(points, key = lambda point: point.x)
+
+        left = xSorted[:2]
+        left.sort(key = lambda point: point.y)
+        tl, bl = left
+
+        right = xSorted[2:]
+        right.sort(key = lambda point: point.y)
+        tr, br = right
+
+        return [
+            tl, tr, br, bl
+        ]
+
+    @classmethod
+    def extractRectangle(cls, mask: np.ndarray) -> 'Rect':
+        contours, _ = cv2.findContours(mask, cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_SIMPLE)
+
+        if len(contours) != 1:
+            raise ValueError("Found more than one contour")
+
+        epsilon = 0.02 * cv2.arcLength(contours[0], True)
+        rectangle = cv2.approxPolyDP(contours[0], epsilon, True)
+
+        points: list[Point2D] = []
+        for point in rectangle:
+            points.append(Point2D(point[0][0], point[0][1]))
+
+        points = cls.sortRectPoints(points)
+        return cls(*points)
+
+
+def generateExtractedImage(
+    image: np.ndarray,
+    segmentationMask: np.ndarray
+) -> Image.Image:
+
+    rgbaImage = Image.fromarray(image).convert("RGBA")
+
+    segmentedImage = np.asarray(rgbaImage) * segmentationMask[..., None]  # reshape segmentationMask for broadcasting
+    segmentedImage = Image.fromarray(segmentedImage)
+
+    rectangle = Rect.extractRectangle(segmentationMask)
+
+    width = rectangle.witdh
+    height = rectangle.height
+
+    transformed = np.float32([[0, 0], [width, 0], [width, height], [0, height]])
+    transformMatrix = cv2.getPerspectiveTransform(rectangle.numpy(), transformed)
+    segmentedImage = Image.fromarray(cv2.warpPerspective(np.array(segmentedImage), transformMatrix, (width, height)))
+
+    alpha = segmentedImage.getchannel("A")
+    bbox = alpha.getbbox()
+    croppedImage = segmentedImage.crop(bbox)
+
+    return croppedImage
+
+
+def processSample(
+    sample: ImageSample,
+    classes: ImageDatasetClasses,
+    excludedClasses: list[str],
+    outputDataset: ImageDataset
+) -> tuple[PILImage, list[CoretexSegmentationInstance]]:
+
+    extractedImagesDir = folder_manager.createTempFolder(sample.name)
+    sampleData = sample.load()
+
+    annotation = sampleData.annotation
+    if annotation is None:
+        raise RuntimeError(f"CTX sample dataset sample id: {sample.id} image doesn't exist!")
+
+    for i, instance in enumerate(annotation.instances):
```

Comment:
If parent class is specified it must be extracted first and opencv perspective transform has to be applied to it before processing other instances. Then extract other instances from the transformed image of the parent class.